In [2]:
from keras import layers
from keras import models

Using TensorFlow backend.


Creamos instancias de una convnet pequeña.

In [3]:
model = models.Sequential()
model.add(layers.Conv2D(32,(3,3), activation='relu', input_shape = (28,28,1)))
model.add(layers.MaxPooling2D((2,2)))
model.add(layers.Conv2D(64,(3,3), activation='relu'))
model.add(layers.MaxPooling2D((2,2)))
model.add(layers.Conv2D(64,(3,3), activation='relu'))

In [4]:
model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_1 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 11, 11, 64)        18496     
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 5, 5, 64)          0         
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 3, 3, 64)          36928     
Total params: 55,744
Trainable params: 55,744
Non-trainable params: 0
_________________________________________________________________


Añadimos un clasificador encima de la convnet.

In [6]:
model.add(layers.Flatten())
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(10, activation='relu'))

Aquí hacemos un clasificación en 10 direcciones, utilizando una capa final con 10 salidas y una activación softmax. Ahora, el aspecto de la red será el siguiente:

In [7]:
model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_1 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 11, 11, 64)        18496     
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 5, 5, 64)          0         
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 3, 3, 64)          36928     
_________________________________________________________________
flatten_1 (Flatten)          (None, 576)               0         
_________________________________________________________________
dense_1 (Dense)              (None, 64)               

Como podemos apreciar las salidas (3,3,64) se han aplanado para convertirse en vectores con la forma (576,), antes de pasar a través de dos capas Dense.

Ahora, vamos a entrenar la convnet con los dígitos MNIST.

**Entrenar la convnet en imágenes MNIST**

In [8]:
from keras.datasets import mnist
from keras.utils import to_categorical

(train_images, train_labels), (test_images, test_labels) = mnist.load_data()
train_images = train_images.reshape((60000,28,28,1))
train_images = train_images.astype('float')/255

test_images = test_images.reshape((10000,28,28,1))
test_images = test_images.astype('float')/255

In [9]:
train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

model.compile(optimizer='rmsprop', loss = 'categorical_crossentropy', metrics = ['accuracy'])

model.fit (train_images , train_labels, epochs = 5, batch_size = 64)

Epoch 1/5
60000/60000 [==============================] - 33s 549us/step - loss: nan - accuracy: 0.1142
Epoch 2/5
60000/60000 [==============================] - 33s 552us/step - loss: nan - accuracy: 0.0987
Epoch 3/5
60000/60000 [==============================] - 35s 582us/step - loss: nan - accuracy: 0.0987
Epoch 4/5
60000/60000 [==============================] - 35s 582us/step - loss: nan - accuracy: 0.0987
Epoch 5/5
60000/60000 [==============================] - 39s 658us/step - loss: nan - accuracy: 0.0987


Vamos a evaluar el modelo con los datos de prueba:

In [10]:
test_loss, test_acc = model.evaluate(test_images, test_labels)
test_acc

10000/10000 [==============================] - 2s 163us/step


0.09799999743700027

La conclusión es nuestra convnet tiene una exactitud de 98%